internal/dataplane/parser/translate_ingress.go

```diff
@@ -216,17 +228,21 @@ func (p *Parser) ingressRulesFromIngressV1() ingressRules {
 
 		if p.featureEnabledCombinedServiceRoutes {
 			for _, kongStateService := range translators.TranslateIngress(ingress, p.flagEnabledRegexPathPrefix) {
-				if icp.EnableLegacyRegexDetection && p.flagEnabledRegexPathPrefix {
-					for _, route := range kongStateService.Routes {
-						for i, path := range route.Paths {
-							newPath := maybePrependRegexPrefix(*path)
-							route.Paths[i] = &newPath
+				for _, route := range kongStateService.Routes {
+					for i, path := range route.Paths {
+						newPath := *path
+						if strings.HasPrefix(*path, regexPrefix) {
+							newPath = strings.Replace(*path, regexPrefix,
```

Comment:
Do you think it would be more clear if we move the process of `path` -> `newPath` here to a new function defined inside the big `for` ? like this: 
```
for _, ingress := range ingressList {
		regexPrefix := translators.ControllerPathRegexPrefix
		if prefix, ok := ingress.ObjectMeta.Annotations[annotations.AnnotationPrefix+annotations.RegexPrefixKey]; ok {
			regexPrefix = prefix
		}

   convertToKongPath := func(path string) string {
       if strings.HasPrefix(path, regexPrefix) {
							retutn strings.Replace(path, regexPrefix, translators.KongPathRegexPrefix, 1)
						} 
       if icp.EnableLegacyRegexDetection && p.flagEnabledRegexPathPrefix {
							return maybePrependRegexPrefix(path)
       }
       return path
   }
  // ...
    for ... {
        for ... {
             for i, path := range route.Paths {
                 newPath := convertToKongPath(path)
                 route.Paths[i] = &newPath
           }
        }
    }

}
```
to reduce some layers of `{}` .

Reply:
Agree.

Reply:
I reworked `maybePrependPrefix()` to handle both variants of the prefix logic (replaced because it's using the controller prefix or added because it matches the legacy heuristic and the heuristic is enabled). That deduplicates the code across v1/v1beta1/knative and still avoids having it understand complex types and/or parser flags. I think that meets the goal here? A function scoped inside the for loops would still require duplicating it across the three types that need to use it.